File: tools/stream_to_static_graph.cpp
/**
 * stream_to_static_graph.cpp
 *
 * Reads a binary edge stream, collects the set of live edges
 * (INSERTs minus DELETEs), canonicalises to (u < v), deduplicates,
 * and writes a binary symmetric CSR file compatible with
 * graph_utils::break_sym_graph_from_bin / read_static_graph_auto.
 *
 * Usage:
 *   ./stream_to_static_graph input_stream.bin output_graph.bin
 */
#include <algorithm>
#include <cstdint>
#include <cstring>
#include <fstream>
#include <iostream>
#include <set>
#include <string>
#include <vector>
#include "binary_graph_stream.h"
#include "types.h"

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <input_stream.bin> <output_graph.bin>\n";
        return 1;
    }
    const std::string input_path = argv[1];
    const std::string output_path = argv[2];

    // Read the stream
    BinaryGraphStream stream(input_path, 1 << 20);  // 1 MB buffer
    uint32_t num_nodes = stream.nodes();
    uint64_t num_ops = stream.edges();

    std::cout << "Stream: " << num_nodes << " nodes, " << num_ops << " operations\n";

    // Collect live edges as a set of canonical (u < v) pairs
    using Edge = std::pair<uint32_t, uint32_t>;
    std::set<Edge> live;

    for (uint64_t i = 0; i < num_ops; ++i) {
        GraphUpdate op = stream.get_edge();
        if (op.type == INSERT) {
            uint32_t u = std::min(op.edge.src, op.edge.dst);
            uint32_t v = std::max(op.edge.src, op.edge.dst);
            if (u != v) live.insert({u, v});
        } else if (op.type == DELETE) {
            uint32_t u = std::min(op.edge.src, op.edge.dst);
            uint32_t v = std::max(op.edge.src, op.edge.dst);
            live.erase({u, v});
        }
        // BREAKPOINT = query, skip
    }

    std::cout << "Live edges after stream: " << live.size() << "\n";

    // Build CSR: adj[u] contains sorted v values where u < v
    std::vector<std::vector<uint32_t>> adj(num_nodes);
    for (const auto& [u, v] : live) {
        adj[u].push_back(v);
    }
    // adj lists are already sorted because std::set iterates in order

    uint64_t num_edges = live.size();

    // Build offsets
    std::vector<uint64_t> offsets(num_nodes + 1);
    offsets[0] = 0;
    for (uint32_t i = 0; i < num_nodes; ++i) {
        offsets[i + 1] = offsets[i] + adj[i].size();
    }

    // Write binary CSR
    uint64_t sizes = (num_nodes + 1) * 8 + num_edges * 4 + 3 * 8;

    std::ofstream out(output_path, std::ios::binary);
    if (!out) {
        std::cerr << "Error: cannot open output file " << output_path << "\n";
        return 1;
    }

    auto write_u64 = [&](uint64_t val) { out.write(reinterpret_cast<const char*>(&val), 8); };
    auto write_u32 = [&](uint32_t val) { out.write(reinterpret_cast<const char*>(&val), 4); };

    // Header
    write_u64(static_cast<uint64_t>(num_nodes));
    write_u64(num_edges);
    write_u64(sizes);

    // Offsets
    for (uint32_t i = 0; i <= num_nodes; ++i) {
        write_u64(offsets[i]);
    }

    // Edge targets
    for (uint32_t i = 0; i < num_nodes; ++i) {
        for (uint32_t v : adj[i]) {
            write_u32(v);
        }
    }

    out.close();
    std::cout << "Wrote " << output_path << ": " << num_nodes << " vertices, "
              << num_edges << " edges, " << sizes << " bytes\n";
    return 0;
}

File: scripts/stream_to_static_graph.py
#!/usr/bin/env python3
"""Convert a binary edge stream to a static graph file (binary symmetric CSR).

Reads a binary stream (.bin), collects every INSERT edge, removes edges that
are later DELETEd, canonicalises to (u < v), deduplicates, and writes a
binary symmetric CSR file compatible with graph_utils::break_sym_graph_from_bin.

Usage:
    python3 scripts/stream_to_static_graph.py input_stream.bin output_graph.bin
"""
import argparse
import os
import struct
import sys

# Stream update types (matches types.h: INSERT=0, DELETE=1, BREAKPOINT=2)
INSERT = 0
DELETE = 1
BREAKPOINT = 2


def read_stream(path: str):
    """Read binary stream and return (num_nodes, set of live edges).

    Each edge is stored as (min(a,b), max(a,b)).
    """
    with open(path, "rb") as f:
        header = f.read(12)
        if len(header) < 12:
            raise ValueError("Stream file too small for header")
        num_nodes = struct.unpack("<I", header[:4])[0]
        num_edges = struct.unpack("<Q", header[4:12])[0]

        edge_size = 1 + 4 + 4  # uint8 + uint32 + uint32
        edges = set()
        buf = f.read()

    if len(buf) < num_edges * edge_size:
        print(
            f"Warning: stream claims {num_edges} edges but file has data for "
            f"{len(buf) // edge_size}",
            file=sys.stderr,
        )

    offset = 0
    count = 0
    while offset + edge_size <= len(buf):
        update_type = buf[offset]
        a = struct.unpack_from("<I", buf, offset + 1)[0]
        b = struct.unpack_from("<I", buf, offset + 5)[0]
        offset += edge_size
        count += 1

        if update_type == INSERT:
            edge = (min(a, b), max(a, b))
            edges.add(edge)
        elif update_type == DELETE:
            edge = (min(a, b), max(a, b))
            edges.discard(edge)
        # BREAKPOINT edges are queries — skip them

    print(f"Read {count} stream operations, {len(edges)} unique live edges", file=sys.stderr)
    return num_nodes, edges


def write_binary_csr(path: str, num_vertices: int, edges: set):
    """Write binary symmetric CSR (break_sym_graph_from_bin format).

    Edges are stored as directed half-edges where src < dst.
    Layout:
        size_t num_vertices
        size_t num_edges (number of half-edges stored)
        size_t sizes     (total file size in bytes)
        uint64_t[num_vertices+1] offsets
        uint32_t[num_edges] edge targets
    """
    # Build adjacency: for each edge (u,v) with u<v, store v under u.
    adj = [[] for _ in range(num_vertices)]
    for u, v in edges:
        if u == v:
            continue
        # Edges are already canonical (u < v)
        adj[u].append(v)

    # Sort each adjacency list
    for i in range(num_vertices):
        adj[i].sort()

    # Build CSR offsets
    num_edges = sum(len(a) for a in adj)
    offsets = [0] * (num_vertices + 1)
    for i in range(num_vertices):
        offsets[i + 1] = offsets[i] + len(adj[i])
    assert offsets[num_vertices] == num_edges

    sizes = (num_vertices + 1) * 8 + num_edges * 4 + 3 * 8

    with open(path, "wb") as f:
        f.write(struct.pack("<Q", num_vertices))
        f.write(struct.pack("<Q", num_edges))
        f.write(struct.pack("<Q", sizes))
        for o in offsets:
            f.write(struct.pack("<Q", o))
        for i in range(num_vertices):
            for v in adj[i]:
                f.write(struct.pack("<I", v))

    print(
        f"Wrote {path}: {num_vertices} vertices, {num_edges} edges, "
        f"{sizes} bytes",
        file=sys.stderr,
    )


def main():
    parser = argparse.ArgumentParser(
        description="Convert a binary edge stream to a static graph file (binary CSR)."
    )
    parser.add_argument("input", help="Input binary stream file (.bin)")
    parser.add_argument("output", help="Output binary CSR graph file")
    args = parser.parse_args()

    if not os.path.isfile(args.input):
        print(f"Error: {args.input} not found", file=sys.stderr)
        sys.exit(1)

    num_nodes, edges = read_stream(args.input)
    write_binary_csr(args.output, num_nodes, edges)


if __name__ == "__main__":
    main()
